### DISCUSSION TITLE
python wildcard S3 path error with parquet

### DISCUSSION THREAD

**Discussion by PeterLietaer** [duckdb/duckdb#3499] [General]
@samansmink, I'm trying the following parquet wildcard query : 
con.execute("select count(*) from parquet_scan('s3://nyc/glob/*.parquet')").fetchall()

I get the following output :

Traceback (most recent call last):
  File "/home/ansibledsi/PycharmProjects/duckdb/main.py", line 78, in <module>
    duckdbtest()
  File "/home/ansibledsi/PycharmProjects/duckdb/main.py", line 42, in duckdbtest
    recordset = con.execute("select count(*) from parquet_scan('s3://nyc/glob/*.parquet')").fetchall()
RuntimeError: HTTP GET error on 'http://192.168.200.25:9000/?encoding-type=url&list-type=2&continuation-token=&prefix=nyc%2Fglob%2F' (HTTP 403)
Forbidden

Process finished with exit code 1

It seems like the url is incorrectly parsed, and all characters starting from the wildcard, are dropped.

Any ideas ?

**Comment by samansmink** [comment]
Hi Peter,

The dropping of all characters after the wildcard is intentional. The way the s3 globbing works is by sending a [ListObjectsV2](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html) API call to S3 to get a list of potential matching for the glob pattern. The final matching then happens in duckdb. 

The HTTP error you're getting seems to be a permission error from the s3 server you're running locally. Does the server you're running locally support the listobjectsv2 call? Also the glob url you're trying looks like its missing a wildcard character?

**Reply by PeterLietaer** [reply]
Hi Sam,

Many thanks for your feedback.
I'm using Minio for S3 storage.  As far as I understand from the documentation, ListObjectsV2 is used by default.
And of course you are right, I forgot a wildcard token in the url.  The correct url should have been 's3://nyc/glob/*.parquet', but unfortunately the problem stays the same.

**Reply by samansmink** [reply]
I haven't tested with minio but it should just work! Could it be that you do not have List permission enabled on the bucket?

**Reply by mskyttner** [reply]
Hi!

I tested with minio locally and while parquet_scan worked fine with individual parquet files, it seems I was not able to get globbing to work:

```
D select count(*) from 's3://demo/*.parquet';
Forbidden
Run Time: real 0.004 user 0.001179 sys 0.002089
Error: HTTP GET error on 'http://minio:9000/?encoding-type=url&list-type=2&continuation-token=&prefix=demo%2F' (HTTP 403)
```

There seems to be a %2F appended to the end of the request url, is that correct?

The `mc ls ducks3/demo` command allows for listing the files in the bucket.

**Reply by samansmink** [reply]
Hi @mskyttner thanks for confirming! %2F is the url encoded slash character, I don't think that should be an issue

I'll try and reproduce this tomorrow using minio

**Reply by mskyttner** [reply]
Hi @samansmink, if you are using docker, there is a reproducible setup using duckdb v 0.3.4 [here](https://github.com/mskyttner/duckdbrest/blob/master/docker-compose.yml) (place the parquet files in s3/demo folder first).

**Comment by samansmink** [comment]
Managed to reproduce it and find out whats wrong.

The issue is that I've only tested the s3 globbing on s3 urls using virtual-hosted–style access: http{s}://{bucket}.{hostname}/{path} For path style access http{s}://{hostname}/{bucket}/{path} this is currently broken. Also switching between path style and virtual host style urls is now implicitly done, it would probably be nicer to make this an explicit option.

An easy workaround would have been to switch to virtual host style access in Minio setting MINIO_DOMAIN, but unfortunately there's currently also another issue where minio doesn't like the empty continuation token that is passed in the listobjectv2 request.

I'll have some time later this week to fix this!

**Comment by AdeelK93** [comment]
Coming over from discussion in #1403 about google cloud storage - I'm getting the same HTTP 403 when globbing a parquet path there too

**Comment by samansmink** [comment]
A fix for s3 globbing was just merged into master: https://github.com/duckdb/duckdb/pull/3629. Both Minio and Google cloud storage using the [interoperability API](https://cloud.google.com/storage/docs/interoperability) should now work with the exception of some key encoding issues with special characters in filenames that I still need to fix.

**Reply by PeterLietaer** [reply]
Hi @samansmink,
thanks for the fix.  I will give it a try asap.

**Reply by mskyttner** [reply]
Hi @samansmink, thanks for the fix. I wanted to try it out using the duckdb CLI against a local minio server and therefore attempted to do a build from main branch for duckdb and to use the httpfs extension from that build. I used this [Dockerfile](https://github.com/mskyttner/duckdbrest/blob/master/Dockerfile.main) to build duckdb CLI and extensions from the main branch. 

When running the statements I used this [init.sql](https://github.com/mskyttner/duckdbrest/blob/master/init.sql) - like before when using v 0.3.4 before globbing support - but slightly changed from before to use the new settings `SET s3_use_ssl=false; SET s3_region='';` (and removing http from the s3_endpoint). 

I guess I have something wrong with the new settings, because running `duckdb -init init.sql` gives this output:

```bash
-- Loading resources from init.sql
Run Time: real 0.003 user 0.002778 sys 0.000000
Error: near line 8: Catalog Error: Table with name s3://demo/*.parquet does not exist!
Did you mean "sqlite_master"?
LINE 1: select count(*) from 's3://demo/*.parquet';                         ^
-- contents of init.sql
load '/extensions/httpfs.duckdb_extension';
set s3_endpoint='minio:9000';
set s3_access_key_id='minioadmin';
set s3_secret_access_key='minioadmin';
set s3_use_ssl=false;
set s3_region='';
.timer on
select count(*) from 's3://demo/*.parquet';
.shell more init.sql
v..1-dev b141a06
Enter ".help" for usage hints.
```
My working "mc alias" is provided through a sysenv MC_HOST_ducks3=http://minioadmin:minioadmin@minio:9000 which is successful in listing parquet files when using "mc ls ducks3/demo".

Do you have any suggestions or pointers? Is it the s3 SET statements that are incorrect or do you think this is an issue with loading the locally built extension?

**Reply by samansmink** [reply]
Hey @mskyttner i think whats happening is that you are compiling duckdb with the DISABLE_BUILTIN_EXTENSIONS flag, which means that none of the extensions are linked into duckdb. This means the parquet extension is also not linked which would cause this error.

Easy solution should be to remove the DISABLE_BUILTIN_EXTENSIONS flag. Then you also do not have to run the LOAD statements as the extensions are linked into the cli during the build.

**Reply by mskyttner** [reply]
@samansmink thanks! Rebuilding with that flag set to 0 and in addition adding "SET s3_url_style='path';" to the previously mentioned init.sql, I too get the local minio s3 parquet globbing to work in the duckdb CLI. I get 

```bash
D set s3_url_style='path';
Run Time: real 0.000 user 0.000063 sys 0.000189
D select count(*) from 's3://demo/*.parquet';
┌──────────────┐
│ count_star() │
├──────────────┤
│ 642152       │
└──────────────┘
Run Time: real 0.013 user 0.004097 sys 0.000113
```
Very nice!

Comparing to the [minio s3 select api](https://docs.min.io/docs/minio-select-api-quickstart-guide.html) a similar query appears to be much faster when running through duckdb, see below a rough comparison (calls being made inside the same container):

```bash
root@f921b524f8ba:/data# time mc sql --recursive --query "select count(*) from S3Object" ducks3/demo/2.parquet
321076

real	0m7.651s
user	0m0.008s
sys	0m0.044s

root@f921b524f8ba:/data# duckdb -init init.sql
-- Loading resources from init.sql
┌──────────────┐
│ count_star() │
├──────────────┤
│ 642152       │
└──────────────┘
Run Time: real 0.009 user 0.001192 sys 0.002384
```

**Reply by samansmink** [reply]
thats great! the timing seems to suggest that the minio client actually does a scan there, duckdb should only read the metadata for a count(*)

**Comment by vtslab** [comment]
Just wanted to let you know that I am anxious to try out #3629 . Any ideas about a release date for 0.3.5? I realize that 0.3.4 is still reasonably new.

**Reply by Alex-Monahan** [reply]
Hello! 

For the moment, have you tried using the latest dev version? Which client do you happen to be using?

We shoot for releases every 1-2 months.

**Comment by vtslab** [comment]
Thanks for pointing me to the dev versions. I use the python client and see that PYPI has dev versions available, so there is no barrier in trying them out!

**Comment by vtslab** [comment]
Using python dev versions with the httpfs extension seems problematic yet:

    con.execute("INSTALL httpfs;")
RuntimeError: IO Error: Failed to download extension http://extensions.duckdb.org/93c64e92f/linux_amd64_gcc4/httpfs.duckdb_extension.gz

I can wait, no problem, I just copy the S3 files to the local file system for now.

**Comment by jacobdanovitch** [comment]
I'm still getting this error with the Python client on v0.51.0. I don't know why it returns a 403 but the issue seems to be the slash following the bucket name; `http://endpoint/bucket/?continuation-token=...` is queried when it seems as though it should be `http://endpoint/bucket?continuation-token=...` being queried instead. 

Is there any workaround?
